### DISCUSSION TITLE
Backend DB: Support multiple hosts for mysql and postgres

### DISCUSSION THREAD

**Discussion by schen59** [grafana/grafana#36966] [Data Sources]
Right now the database config seems only support connecting to a single host. Can we have some support to connecting a database cluster? Basically we can specify the list of host in the cluster in the config file, when doing database request either we can randomly pick up a host to connect or have some basic load balancing.

**Comment by torkelo** [comment]
is there not loadbalancers for mysql or postgres clusters that can do that?

**Comment by schen59** [comment]
Ya, there is. Problem is that we need extra setup for the load balancer. In addition, seems load balancer will make the cluster public facing? Like pymongo package, it supports connect to a list of hosts. Not sure if this is a valid feature request. Personally, I think it is a nice to have feature.

**Comment by torkelo** [comment]
not sure why it would need to be public facing in order to use a load balancer

**Comment by nvtkaszpir** [comment]
I'm affraid this feature is out of scope of grafana.

First, you want to limit to read only, which is easy - then yes, it is doable, but still you need to implemetnt multiple host handling in grafana (backend proxy and gui I guess). And still you would have to check which hosts are healthy and then shuffle among the sent request to them to load balance it.

If you want to also use it for writing, then stuff gets more complicated - you would have to define a host to send all writes to. If you would like to define a list of hosts to send writes to, then things gets VERY complicated on the grafana side.

I guess If someone has such request, then should actually set up haporxy and configure grafana to connect to it - haproxy can redirect plain tcp connections to the backend servers for example SQL.
Such setup is very easy to implement (something like 20lines?) and does not have to be exposed publicly. Setting up haproxy on the same node where grafana is hosted  and bound to 127.0.0.1 is really easy.

**Comment by adeverteuil** [comment]
For reference, here is what I found in the PostgreSQL docs about connection strings and specifying multiple hosts.

https://www.postgresql.org/docs/10/libpq-connect.html#LIBPQ-CONNSTRING

> Several libpq functions parse a user-specified string to obtain connection parameters. There are two accepted formats for these strings: plain `keyword = value` strings and URIs. URIs generally follow `RFC 3986`, except that multi-host connection strings are allowed as further described below.

https://www.postgresql.org/docs/10/libpq-connect.html#LIBPQ-MULTIPLE-HOSTS

> It is possible to specify multiple host components, each with an optional port component, in a single URI. A URI of the form `postgresql://host1:port1,host2:port2,host3:port3/` is equivalent to a connection string of the form `host=host1,host2,host3 port=port1,port2,port3`. Each host will be tried in turn until a connection is successfully established.

There are two connection string formats (here with multiple hosts example):

```
postgresql://host1:123,host2:456/somedb?target_session_attrs=any&application_name=myapp
host=host1,host2 port=5432 dbname=mydb connect_timeout=10
```

Grafana accepts `postgres://` but not `postgresql://` although both are valid according to PostgreSQL docs.

> The URI scheme designator can be either postgresql:// or postgres://.

Testing this in Grafana configuration, it looks like these are not parsed according to the above. Grafana fails to start if I use multiple hosts syntax.

```
[database]
url = postgres://host1:123,host2:456/somedb?target_session_attrs=any&application_name=myapp
```

```
EROR[05-04|08:41:59] Server shutdown logger=server reason="Service init failed: Fail to connect to database: Invalid host specifier 'host1:123,host2:456': net.SplitHostPort failed for 'host1:123,host2:456': address host1:123,host2:456: too many colons in address"
```

Does the Go client library for PostgreSQL used by Grafana support this? If so, then Grafana is doing something with the arguments and is not passing them correctly to the client library.

**Comment by marefr** [comment]
@adeverteuil can you try using the `connection_string` setting instead of `url`?

**Comment by germm** [comment]
The lib/pq PostgreSQL driver can only connect to a single host (see [libpq 683](https://github.com/lib/pq/issues/683)) . A [pull request](https://github.com/lib/pq/pull/714) to fix this issue is pending for several years.

As the lib/pq driver is in maintenance mode, this feature is unlikely to be added. lib/pq recommends to use [pgx](https://github.com/jackc/pgx) as a replacement.

Do you have plans to migrate Grafana to the pgx PostgreSQL Driver?

**Comment by aknuds1** [comment]
Hadn't heard of pgx @germm - thanks for the tip.

**Comment by mfvanek** [comment]
We are highly interested in this feature

**Comment by nvtkaszpir** [comment]
As I stated over 5 years ago, this is rather out of scope of the grafana.

This can be solved with:
- for mysql - proxysql
- for posgres - pgbounce or pgpool, or something like stolon-proxy

or just create a separate read-only host and point grafana at it - this is the easiest and the safest solution, and for HA you can just hide multiple read-only replicas behind something like haproxy/nginx.

**Reply by siepkes** [reply]
All the alternatives you list are active components. Meaning they can fail and have to be setup, maintained, etc. I've personally used Stolon, but Stolon also requires a coordination cluster such as Zookeeper or Consul for coordination.

Most other PostgreSQL client drivers (JDBC, Python, Perl, etc.) simply allow you to specify multiple DB hosts in the URL. You can also configure them to always connect to a writeable host. For example with JDBC it looks like `jdbc:postgresql://postgresql-1.foo.com,postgresql-2.foo.com,postgresql-3.foo.com/postgres?targetServerType=primary&hostRecheckSeconds=5&currentSchema=fooschema`. This is way less maintenance then using something like Stolon.

**Reply by krisavi** [reply]
Only thing grafana side should do is support whole pg library functionality and not limit it artificially. Currently unfortunately the pg library used is already long time in maintenance mode and shifting to another library would not be too much work, would support HA solution built into libpq and avoiding using additional links in the chain that could fail. 

As for something being in grafana's scope I would say that using maintained libraries would be something that should be in the scope. lib/pq seems to be in maintenance and not support multi-host, however jackc/pgx and go-pg/pg both seem to support this requested functionality and both libraries are actively maintained.

@nvtkaszpir How would read-only host help if grafana stores it's config in database? Using PG as datasource, yeah you could point to read-only node, but still there can be cases where that specific node needs some maintenance or issues connecting to it for some reason while other nodes still available, but as for grafana, then you see it as missing datasource.

Main issue is that currently when there is DB switchover from one node to another, then without separate tool in between to direct traffic to correct host, you have to change grafana.ini and with it heavily limiting HA capabilities of grafana. 

I cannot speak for mysql solutions.

Pgbouncer, that was mentioned will not help to solve this problem in any way. pgPool-II yeah, but is not easy to set up. If you happen to use patroni for cluster management, then Stolon is not going to work, as Stolon is not just proxy.

**Comment by suntala** [comment]
Closing this discussion as it is stale. Please open as an issue if you'd like to make a feature request.

**Reply by siepkes** [reply]
For anyone stumbling across this discussion: That would be this issue: https://github.com/grafana/grafana/issues/39422
